service/history/workflow/update/registry_test.go

```diff
@@ -317,47 +317,81 @@ func TestFindOrCreate(t *testing.T) {
 	})
```

Comment:
> I may miss something, but none of those tests seems to check completedCount?

I think it's better for these tests to assert the behavior (ie rate limited or not) instead of checking the internal state directly.

Reply:
> since this is unit tests, I will recommend to test that remover function directly

I doesn't say unit test anywhere ;) it's really an integration test ...

Reply:
You are using the knowledge of internal state (counter will not increase under some conditions) to test the behavior of different abstraction (rate limiter). 
This is trivial case, but still a bad practice.
Integration test exist for a different purpose - to make sure they work when combined. You may add integration test to check that rate limit is working with your component.
But you shouldn't use it to test internal behavior and assumptions of the code that change the counter. 

Now if you change the logic this test will fail, but integration test shouldn't fail in this case - the contract between those two components is not violated.

Reply:
(not blocking since this is discussion point)